### DISCUSSION TITLE
Enumeration value in JSON-LD without @context

### DISCUSSION THREAD

**Discussion by KalleOlaviNiemitalo** [schemaorg/schemaorg#3355] [Q&A]
What's the right way to represent a schema.org enumeration value in JSON-LD without using `@context`?  Does it follow from some specification?

* RDF label:
  ```JSON
  {
      "@type": "https://schema.org/CreateAction",
      "https://schema.org/actionStatus": "CompletedActionStatus"
  }
  ```
* Value object:
  ```JSON
  {
      "@type": "https://schema.org/CreateAction",
      "https://schema.org/actionStatus": {
          "@value": "CompletedActionStatus",
          "@type": "https://schema.org/ActionStatusType"
      }
  }
  ```
* `@id`:
  ```JSON
  {
      "@type": "https://schema.org/CreateAction",
      "https://schema.org/actionStatus": {
          "@id": "https://schema.org/CompletedActionStatus"
      }
  }
  ```
## Analysis

The <https://schema.org/actionStatus> property ([JSON-LD](https://github.com/schemaorg/schemaorg/blob/12e9bbb7090436748e5ee393ee5fafd395a43bbf/data/releases/22.0/schemaorg-current-https.jsonld#L10003-L10014)) has a range that includes <https://schema.org/ActionStatusType> ([JSON-LD](https://github.com/schemaorg/schemaorg/blob/12e9bbb7090436748e5ee393ee5fafd395a43bbf/data/releases/22.0/schemaorg-current-https.jsonld#L1515-L1523)), which is a subclass of <https://schema.org/StatusEnumeration> ([JSON-LD](https://github.com/schemaorg/schemaorg/blob/12e9bbb7090436748e5ee393ee5fafd395a43bbf/data/releases/22.0/schemaorg-current-https.jsonld#L6307-L6318)) and thus also <https://schema.org/Enumeration> ([JSON-LD](https://github.com/schemaorg/schemaorg/blob/12e9bbb7090436748e5ee393ee5fafd395a43bbf/data/releases/22.0/schemaorg-current-https.jsonld#L13579-L13587)).

The <https://schema.org/CompletedActionStatus> enumeration value ([JSON-LD](https://github.com/schemaorg/schemaorg/blob/12e9bbb7090436748e5ee393ee5fafd395a43bbf/data/releases/22.0/schemaorg-current-https.jsonld#L29623-L29628)) is an instance of <https://schema.org/ActionStatusType> and has "CompletedActionStatus" as its label.

If I understand correctly, the <https://schema.org/CompletedActionStatus> node would be a suitable value for the <https://schema.org/actionStatus> property.  The string "CompletedActionStatus" would not be suitable as is, and would have to be coerced.  However, it is not clear to me whether this coercion would be done as part of converting from JSON-LD to tuples, or in a subsequent processing step. It is also not clear whether the coercion would be based on the <http://www.w3.org/2000/01/rdf-schema#label> property of the enumeration members, on the path components of the URLs, or on hardcoded knowledge about the <https://schema.org/actionStatus> property and the <https://schema.org/ActionStatusType> enumeration type.

## If `"@context": "https://schema.org/"` were used

The JSON-LD context at <https://schema.org/docs/jsonldcontext.jsonld> declares "actionStatus" as meaning <http://schema.org/actionStatus>, but does not specify a `@type` for coercing the property value:

<https://github.com/schemaorg/schemaorg/blob/12e9bbb7090436748e5ee393ee5fafd395a43bbf/data/releases/22.0/schemaorgcontext.jsonld#L24> <https://github.com/schemaorg/schemaorg/blob/12e9bbb7090436748e5ee393ee5fafd395a43bbf/data/releases/22.0/schemaorgcontext.jsonld#L1433>

The context also declares "CompletedActionStatus" as meaning <http://schema.org/CompletedActionStatus>:

<https://github.com/schemaorg/schemaorg/blob/12e9bbb7090436748e5ee393ee5fafd395a43bbf/data/releases/22.0/schemaorgcontext.jsonld#L250>

If the context specified `"@type": "@vocab"` for the <http://schema.org/actionStatus> property, then that would cause the JSON-LD consumer to treat the property value "CompletedActionStatus" as a term and expand it to <http://schema.org/CompletedActionStatus>.  The context doesn't currently do so, though.  And of course this cannot help when the JSON-LD document does not use `@context`.

**Comment by WeaverStever** [comment]
The schema [validator](https://validator.schema.org/) seems to extract "CompletedActionStatus" to http://schema.org/CompletedActionStatus

~~~~
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "ListenAction",
  "agent": {
    "@type": "Person",
    "name": "John"
  },
  "object": {
    "@type": "MusicGroup",
    "name": "Pink!"
  },
  "participant": {
    "@type": "Person",
    "name": "Steve"
  },
  "location": {
    "@type": "Residence",
    "name": "Ann's apartment"
  },
  "instrument": {
    "@type": "Product",
    "name": "iPod"
  },
  "actionStatus": "CompletedActionStatus"
}
</script>
~~~~

**Reply by KalleOlaviNiemitalo** [reply]
That validator seems to have some built-in knowledge about the <http://schema.org/actionStatus> property even when `@context` is not used.  It coerces "CompletedActionStatus" to <http://schema.org/CompletedActionStatus> from the following input, too:

```JSON
{
  "@type": "http://schema.org/ListenAction",
  "http://schema.org/agent": {
    "@type": "http://schema.org/Person",
    "name": "John"
  },
  "http://schema.org/object": {
    "@type": "http://schema.org/MusicGroup",
    "name": "Pink!"
  },
  "http://schema.org/participant": {
    "@type": "http://schema.org/Person",
    "name": "Steve"
  },
  "http://schema.org/location": {
    "@type": "http://schema.org/Residence",
    "name": "Ann's apartment"
  },
  "http://schema.org/instrument": {
    "@type": "http://schema.org/Product",
    "name": "iPod"
  },
  "http://schema.org/actionStatus": "CompletedActionStatus"
}
```

AFAICT, this coercion is not part of the JSON-LD specification.  But I'd like to understand whether it is

- standardised at schema.org,
- standardised as part of RDF, or
- ad-hoc implemented by individual schema.org consumers in order to make sense of web pages that they crawl.

**Comment by gkellogg** [comment]
JSON-LD does not perform any vocabulary processing, so the fact that the range of `actionStatus` is `Action` is not used when expanding the JSON-LD, or turning it into RDF. The only way this is done is if the term definition for `actionStatus` in an included context is defined with `"@type": "@id"` (or `"@type": "@vocab"`), which it apparently is not, thus "CompletedActionStatus" must be interpreted as a literal value.

This does not mean that other tools can't use the stated range to infer something different, but the Linter, for one, will not. Generally, schema.org has a high tolerance for accepting strings where an IRI (or object) might be expected to the point where properties are often defined to have a textual range, so the schema.org context is not overly aggressive in using `@type` in a term definition. Certainly, it would not be hard to infer that the literal value "CompletedActionStatus" was really intended to represent `http://schema.org/CompletedActionStatus` using the available information.

**Comment by WeaverStever** [comment]
This look relevant...
https://jazz.net/wiki/bin/view/LinkedData/UseUrisForEnums

(Edited: for explaining enumerations in RDF)

**Comment by rrlevering** [comment]
For the record, the schema.org validator is essentially doing the non-standardised third option which is part of a "coercion" layer (read "Anything non-standard that is common on the web")  We have a copy of the schema definition from schema.org and we normalize "<Value>" -> "https://schema.org/<Value>" if it's a valid enum value for any enumeration in the range.

**Comment by danbri** [comment]
Let’s call it normalization or canonicalization, to avoid
misunderstandings. I prefer the former as the same data might be normalized
differently for different consuming environments. In practice all schema.org
consumers need to do quite a lot of this…
